# Accident Severity with Artificial Neural Network (ANN)

We load the ml_ready.csv file which we got from our decision tree version of the prediction. (prediction/supervised/preprocess_data.py)

In [1]:
from ann import train
import pandas as pd

In [2]:
df = pd.read_csv('../data/ml_ready.csv')
df

Accident_Severity      Time  Age_of_Vehicle  \
0                      0.0  0.733658       -0.925537   
1                      0.0  0.009736       -0.474976   
2                      0.0  0.440890       -0.700257   
3                      0.0  0.884562        0.651424   
4                      0.0  0.527816       -1.376097   
...                    ...       ...             ...   
1892446                0.0  0.280946        0.200864   
1892447                0.0  0.698192        1.552544   
1892448                0.0  0.698192       -0.024416   
1892449                0.0  0.833797        1.327264   
1892450                0.0  0.520862       -0.474976   

         Light_Conditions_Darkness - lighting unknown  \
0                                               False   
1                                               False   
2                                               False   
3                                                True   
4                                               False   
...                                               ...   
1892446                                         False   
1892447                                         False   
1892448                                         False   
1892449                                         False   
1892450                                         False   

         Light_Conditions_Darkness - lights lit  \
0                                          True   
1                                          True   
2                                         False   
3                                         False   
4                                         False   
...                                         ...   
1892446                                    True   
1892447                                   False   
1892448                                   False   
1892449                                    True   
1892450                                   False   

         Light_Conditions_Darkness - lights unlit  \
0                                           False   
1                                           False   
2                                           False   
3                                           False   
4                                           False   
...                                           ...   
1892446                                     False   
1892447                                     False   
1892448                                     False   
1892449                                     False   
1892450                                     False   

         Light_Conditions_Darkness - no lighting  Light_Conditions_Daylight  \
0                                          False                      False   
1                                          False                      False   
2                                          False                       True   
3                                          False                      False   
4                                          False                       True   
...                                          ...                        ...   
1892446                                    False                      False   
1892447                                    False                       True   
1892448                                    False                       True   
1892449                                    False                      False   
1892450                                    False                       True   

         Weather_Conditions_Fine + high winds  \
0                                       False   
1                                       False   
2                                       False   
3                                       False   
4                                       False   
...                                       ...   
1892446                                 False   
1892447                                 False   
1892448    

In [3]:
train.train(df)

C:\Users\Jonas\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8581 - loss: 0.4334 - val_accuracy: 0.8586 - val_loss: 0.4288
Epoch 2/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8579 - loss: 0.4296 - val_accuracy: 0.8586 - val_loss: 0.4284
Epoch 3/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8576 - loss: 0.4300 - val_accuracy: 0.8586 - val_loss: 0.4282
Epoch 4/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8568 - loss: 0.4311 - val_accuracy: 0.8586 - val_loss: 0.4281
Epoch 5/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8578 - loss: 0.4292 - val_accuracy: 0.8586 - val_loss: 0.4276
Epoch 6/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8574 - loss: 0.4297 - val_accuracy: 0.8586 - val_loss: 0.4281
Epoch 7/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 0.8576 - loss: 0.4297 - val_accuracy: 0.8586 - val_loss: 0.4295
Epoch 8/10
37849/37849 ━━━━━━━━━━━━━━━━━━━━ 50s 1ms/step - accuracy: 

# Evaluation

* Class 0 = Slight severity
* Class 1 = Serious severity
* Class 2 = Fatal severity

## ROC AUC Scores
* AUC Score (One-vs-Rest): 0.701: This score represents the average AUC of each class against all other classes. An AUC of 0.701 suggests that the model has moderate discriminatory ability overall.
* AUC Score (One-vs-One): 0.701: This score represents the average AUC for each pair of classes. The same score of 0.701 here suggests that the model performs consistently when distinguishing between any two classes.

## Classification Report
### Precision:
Precision is the ratio of true positive predictions to the total predicted positives. High precision indicates that the model does not have many false positives.
* Class 0: Precision is 0.86, meaning that 86% of the instances predicted as Class 0 are correct.
* Class 1: Precision is 0.29, indicating a high number of false positives.
* Class 2: Precision is 1.00, but this is misleading because it is based on zero actual predictions (see confusion matrix).

### Recall:
Recall is the ratio of true positive predictions to the total actual positives. High recall indicates that the model captures most of the positive instances.

* Class 0: Recall is 1.00, meaning the model correctly identifies all instances of Class 0.
* Class 1: Recall is 0.00, indicating that the model fails to identify Class 1 instances.
* Class 2: Recall is 0.00, meaning the model fails to identify Class 2 instances.

### F1-Score: 
The F1-score is the harmonic mean of precision and recall. It balances the two metrics.

* Class 0: F1-score is 0.92, indicating strong performance.
* Class 1: F1-score is 0.00, indicating poor performance.
* Class 2: F1-score is 0.00, indicating poor performance.

### Support:
Support is the number of actual occurrences of each class in the test set.

* Class 0: 325002 instances.
* Class 1: 48379 instances.
* Class 2: 5110 instances.

## Confusion Matrix
The confusion matrix provides a more detailed look at how the model's predictions are distributed across the classes.

* Class 0 (Majority Class): Almost all instances (324977 out of 325002) are correctly classified, with only 25 misclassified as Class 1.
* Class 1: The model almost entirely fails to identify Class 1 instances, with 48369 misclassified as Class 0 and only 10 correctly classified.
* Class 2: The model completely fails to identify Class 2 instances, with all 5110 misclassified as Class 0.]]


## Summary
### Strengths:
The model is very good at identifying the majority class (Class 0), as indicated by the high precision, recall, and F1-score for this class.
### Weaknesses:
The model performs very poorly on the minority classes (Class 1 and Class 2), with almost no correct predictions for these classes. This indicates a significant class imbalance issue where the model is biased towards the majority class.

## Things we should have done different:

### Address Class Imbalance:
We could consider techniques like oversampling the minority classes, undersampling the majority class, or using class weights to balance the training process.
### Model Tuning:
Experiment with different architectures, hyperparameters, or alternative models that might handle class imbalance better.
